modules/mastodon/index.js

```diff
@@ -0,0 +1,320 @@
+  return {'instances': require('./default_instances.json')};
+}
+
+function writeConfig() {
+  bot.writeDataFile('config.json', JSON.stringify(moduleConfig), function(err) {
+    if (err)
+      log.error('error writing mastodon module data: ' + err);
+  });
+}
+
+function checkHostname(hostname) {
+  if (moduleConfig.blocks && moduleConfig.blocks.includes(hostname))
+    return false;
+
+  return moduleConfig.instances && moduleConfig.instances.includes(hostname);
+}
+
+/* TODO: Add support for validating GNU social? */
+/* TODO: Don't check instances more than once a day. */
+function forceCheckHostname(url, fn) {
+  robotRequest(
+    { url: urlMod.resolve(url, '/api/v1/instance')},
+    (error, response, body) => {
+      if (error || response.statusCode != 200)
+        return fn(false);
+
+      try {
+        body = JSON.parse(body);
+      } catch (e) {
+        return fn(false);
+      }
+
+      if (!body.uri)
+        return fn(false);
+
+      fn(true);
+      log.info(`adding ${response.request.host} to instance list`);
+      if (moduleConfig.instances) {
+        moduleConfig.instances.push(response.request.host);
+      } else {
+        moduleConfig.instances = [response.request.host];
+      }
+      writeConfig();
+    }
+  );
+}
+
+/*
+ * fetchResource only attempts to fetch the URL given as a particular content
+ * type. This is less than what a full ActivityPub-implementing client needs to
+ * handle; following Link headers and <link> HTML tags of the correct content
+ * types would also be required.
+ */
+function fetchResource(url, fn) {
+  robotRequest({
+    url: url,
+    headers: { 'Accept': 'application/ld+json; profile="https://www.w3.org/ns/activitystreams", application/activity+json' },
+  }, (error, response, body) => {
+    if (error) {
+      fn(error);
+    } else if (response.statusCode != 200) {
+      fn(new Error('response status code is ' + response.statusCode));
+    } else if (!['application/ld+json', 'application/activity+json'].includes(
+      contentType.parse(response.headers['content-type']).type
+    )) {
+      fn(new Error('resource is not an ActivityPub record'));
+    } else {
+      try {
+        var record = JSON.parse(body);
+        if (!['Note', 'Person'].includes(record.type)) {
+          fn(new Error('record is not a Note or Person'));
+        } else {
+          fn(null, record);
+        }
+      } catch (e) {
+        fn(e);
+      }
+    }
+  });
+}
+
+function htmlToText(s) {
+  return htmlToTextMod.fromString(s, {
+    wordwrap: null,
+    ignoreHref: true,
+  });
+}
+
+function formatRecord(record, fn) {
+  if (record.type == 'Note') {
+    fetchResource(record.attributedTo, (error, userRecord) => {
+      var str = '';
+      if (error) {
+        str += '(error getting user): ';
+      } else {
+        str += `${userRecord.name} (@${userRecord.preferredUsername}): `;
+      }
+      if (record.sensitive) {
+        str += sprintf('[content warning] %s', record.summary);
```

Comment:
Another sprintf, and if we use `${templating}` for this one, we can remove the `sprintf` dependency